abstract/encoding.go

```diff
@@ -170,13 +193,31 @@ func (de *decoder) value(v reflect.Value, depth int) error {
 			}
 		}
 
-	case reflect.Slice, reflect.Array:
+	case reflect.Slice:
+		if v.IsNil() {
+			panic("slices must be initialized to correct length before decoding")
+		}
+		fallthrough
+	case reflect.Array:
 		l := v.Len()
 		for i := 0; i < l; i++ {
 			if err = de.value(v.Index(i), depth+1); err != nil {
 				return err
 			}
 		}
+
+	case reflect.Int:
```

Comment:
Right.  It sounds here like you're arguing for not supporting the size-unspecified 'int' type in the binary/rigid encoding at all, right?  At least that's the only solution I can see here that would avoid any risk of those types of bugs.  Or are you proposing something else?

On the other hand, I might argue that there's nothing specific to network wire-encodings in the general class of bug risk you're pointing out.  I might just as well have an 'int' variable that's used only purely locally in my program, which counts the same thing (e.g., number of milliseconds since the first something), and again works fine on a 64-bit machine but dies on a 32-bit machine.  This is a risk we run all the time when we use limited-size int types whose size depends on the platform, and it's certainly a problem, but not one that Go tries to deal with.  

In fact I'm not sure I really like the fact that Go even has a size-unspecified int type at all - that's one way in which I think Java was perhaps better (where 'int' is always 32-bits and 'long' is always 64-bits).  But I do kinda understand their motivations for sticking with this somewhat unfortunate C-ism.  Tastes and philosophies differ here.  But again that doesn't seem like anything specific to a wire format; these considerations and tradeoffs apply to all code written in the language.